linera-faucet/server/src/database.rs

```diff
@@ -0,0 +1,283 @@
+
+        if end_height == BlockHeight::ZERO {
+            info!("Chain is empty, no synchronization needed");
+            return Ok(BTreeMap::new());
+        }
+
+        let mut height_to_hash = BTreeMap::new();
+        let mut current_hash = info.block_hash;
+
+        // Traverse backwards to build the height -> hash mapping and find sync point
+        while let Some(hash) = current_hash {
+            let certificate = client
+                .storage_client()
+                .read_certificate(hash)
+                .await?
+                .ok_or_else(|| anyhow::anyhow!("Certificate not found for hash {}", hash))?;
+            let current_height = certificate.block().header.height;
+
+            // Check if this block's chains are already in our database
+            let chains_in_block = self.extract_opened_chains(&certificate).await?;
+
+            if !chains_in_block.is_empty() {
+                let mut all_chains_exist = true;
+                for (owner, _chain_id) in &chains_in_block {
+                    if self.get_chain_id(owner).await?.is_none() {
+                        all_chains_exist = false;
+                        break;
+                    }
+                }
+
+                if all_chains_exist {
+                    // All chains from this block are already in the database.
+                    break;
+                }
+            }
+
+            // Add to our height->hash map
+            height_to_hash.insert(current_height, hash);
+
+            // Move to the previous block
+            current_hash = certificate.block().header.previous_block_hash;
+        }
+
+        Ok(height_to_hash)
+    }
+
+    /// Syncs the database using a pre-built height->hash map.
+    async fn sync_forward_with_map<E>(
+        &self,
+        client: &ChainClient<E>,
+        height_to_hash: BTreeMap<BlockHeight, CryptoHash>,
+    ) -> anyhow::Result<()>
+    where
+        E: linera_core::Environment,
+        E::Storage: Storage,
+    {
+        if height_to_hash.is_empty() {
+            return Ok(());
+        }
+
+        // Process blocks in chronological order (forward)
+        for (height, hash) in height_to_hash {
+            let certificate = client
+                .storage_client()
+                .read_certificate(hash)
+                .await?
+                .ok_or_else(|| anyhow::anyhow!("Certificate not found for hash {}", hash))?;
+
+            let chains_to_store = self.extract_opened_chains(&certificate).await?;
+
+            if !chains_to_store.is_empty() {
+                info!(
+                    "Processing block at height {height} with {} new chains",
+                    chains_to_store.len()
+                );
+                self.store_chains_batch(chains_to_store).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Extracts OpenChain operations from a certificate and returns (owner, chain_id) pairs.
+    async fn extract_opened_chains(
```

Comment:
from a purely software engineering perspective: I don't think this should be a method on a database instance

Reply:
It also duplicates some code from `lib.rs`, and its logic is questionable: It shouldn't bother going through the transactions at all, and just use the descriptions. I'll work on a follow-up PR…